clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java

```diff
@@ -1199,6 +1199,168 @@
 
     public static final Schema[] INIT_PRODUCER_ID_RESPONSE = new Schema[] {INIT_PRODUCER_ID_RESPONSE_V0};
 
+    public static final Schema ADD_PARTITIONS_TO_TXN_REQUEST_V0 = new Schema(
+            new Field("transactional_id",
+                    STRING,
+                    "The transactional id corresponding to the transaction."),
+            new Field("pid",
+                    INT64,
+                    "Current PID in use by the transactional id"),
+            new Field("epoch",
+                    INT16,
+                    "Current epoch associated with the PID"),
+            new Field("topic_partitions",
+                    new ArrayOf(new Schema(
+                            new Field("topic", STRING),
+                            new Field("partitions", new ArrayOf(INT32)))),
+                    "The partitions to add to the transaction")
+    );
+    public static final Schema ADD_PARTITIONS_TO_TXN_RESPONSE_V0 = new Schema(
+            new Field("error_code",
+                    INT16,
+                    "An integer error code.")
+    );
+
+    public static final Schema[] ADD_PARTITIONS_TO_TXN_REQUEST = new Schema[] {ADD_PARTITIONS_TO_TXN_REQUEST_V0};
+    public static final Schema[] ADD_PARTITIONS_TO_TXN_RESPONSE = new Schema[] {ADD_PARTITIONS_TO_TXN_RESPONSE_V0};
+
+    public static final Schema ADD_OFFSETS_TO_TXN_REQUEST_V0 = new Schema(
+            new Field("transactional_id",
+                    STRING,
+                    "The transactional id corresponding to the transaction."),
+            new Field("pid",
+                    INT64,
+                    "Current PID in use by the transactional id"),
+            new Field("epoch",
+                    INT16,
+                    "Current epoch associated with the PID"),
+            new Field("consumer_group_id",
+                    STRING,
+                    "Consumer group id whose offsets should be included in the transaction")
+    );
+    public static final Schema ADD_OFFSETS_TO_TXN_RESPONSE_V0 = new Schema(
+            new Field("error_code",
+                    INT16,
+                    "An integer error code.")
+    );
+
+    public static final Schema[] ADD_OFFSETS_TO_TXN_REQUEST = new Schema[] {ADD_OFFSETS_TO_TXN_REQUEST_V0};
+    public static final Schema[] ADD_OFFSETS_TO_TXN_RESPONSE = new Schema[] {ADD_OFFSETS_TO_TXN_RESPONSE_V0};
+
+    public static final Schema END_TXN_REQUEST_V0 = new Schema(
+            new Field("transactional_id",
+                    STRING,
+                    "The transactional id corresponding to the transaction."),
+            new Field("pid",
+                    INT64,
+                    "Current PID in use by the transactional id"),
+            new Field("epoch",
+                    INT16,
+                    "Current epoch associated with the PID"),
+            new Field("transaction_result",
+                    INT8,
```

Comment:
Are we actually sending `UNKNOWN` in the wire? If so, that settles it.

In terms of future flexibility, are we saying that we want to have the ability to add new values without bumping the protocol version? If not, then the flexibility argument doesn't seem to matter (once we bump the protocol version, we can change any type as we please).